server/data_common/matrix_loader.py

```diff
@@ -0,0 +1,173 @@
+
+    # FIXME:  If the actual dataset is changed.  E.g. a new set of datafiles replaces an existing set,
+    # then the cache will not react to this.   Ideally this would invalidate the cache.  One solution is
+    # to keep a small metadata file associated with each dataset, which contains versioning information.
+    # When the dataset is accessed, the current version can be compared with the cached version, and if
+    # there is a mismatch, then the cache can be refreshed.
+
+    def __init__(self):
+        # key is location, value is tuple of (MatrixDataCacheItem, last_accessed)
+        self.datasets = {}
+        self.lock = threading.Lock()
+
+    @contextmanager
+    def data_adaptor(self, location, app_config):
+        # create a loader for to this location if it does not already exist
+        with self.lock:
+            value = self.datasets.get(location)
+            if value is not None:
+                cache_item = value[0]
+                last_accessed = time.time()
+                self.datasets[location] = (cache_item, last_accessed)
+            else:
+                while True:
+                    # find the last access times for each loader
+                    items = list(self.datasets.items())
+                    sorted(items, key=lambda x: x[1][1])
+                    if len(items) < self.MAX_CACHED:
+                        break
+
+                    # close the least recently used loader
+                    oldest = items[0]
+                    oldest_cache = oldest[1][0]
+                    oldest_key = oldest[0]
+                    oldest_cache.delete()
+                    del self.datasets[oldest_key]
+
+                last_accessed = time.time()
+                loader = MatrixDataLoader(location)
+                cache_item = MatrixDataCacheItem(loader)
+                self.datasets[location] = (cache_item, last_accessed)
+        try:
+            data_adaptor = cache_item.acquire(app_config)
+            yield data_adaptor
+        finally:
+            cache_item.release()
+
+
+class MatrixDataType(Enum):
+    H5AD = "h5ad"
+    TILEDB = "tiledb"
+    UNKNOWN = "unknown"
+
+
+class MatrixDataLoader(object):
+
+    def __init__(self, location, etype=None):
+        self.location = location
+        if etype is None:
+            self.etype = self.matrix_data_type()
+        else:
+            self.etype = etype
+        self.matrix_type = None
+        if self.etype == MatrixDataType.H5AD:
+            from server.data_scanpy.scanpy_adaptor import ScanpyAdaptor
+            self.matrix_type = ScanpyAdaptor
+        elif self.etype == MatrixDataType.TILEDB:
+            from server.data_tiledb.tiledb_adaptor import TileDbAdaptor
+            self.matrix_type = TileDbAdaptor
+
+    def matrix_data_type(self):
+        if self.location.endswith(".h5ad"):
+            return MatrixDataType.H5AD
+        elif ".cxg" in self.location:
+            return MatrixDataType.TILEDB
```

Comment:
nit -- the actual datatype is a CXG, not TILEDB.   CXG is built upon TileDB for sure, but wraps a bunch of other semantics around it (which we depend upon).

In a similar way, H5AD is built upon HDF5, but is not equivalent.

Recommend a global search/replace of MatrixDataType.TILEDB with MatrixDataType.CXG

Reply:
How about also changing tiledb_adaptor to cxg_adaptor ?

Reply:
yes, I like it.

Reply:
done.